In [1]:
import pandas as pd
import numpy as np
import scipy as sp

import statsmodels.api as sm
import statsmodels.formula.api as smf
import statsmodels.stats.api as sms
import sklearn as sk

import datetime as dt
from patsy import dmatrix

import matplotlib.pylab as plt
from mpl_toolkits.mplot3d import Axes3D

import seaborn as sns
sns.set()
sns.set_style("whitegrid")
sns.set_color_codes(palette="muted")

%matplotlib inline
%config InlineBackend.figure_formats = {'png', 'retina'}

C:\Users\ledes\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [4]:
from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold

In [2]:
taxi = pd.read_csv("latest_working_file.csv")

In [3]:
taxi.describe()

Unnamed: 0  passenger_count          dist       holiday  \
count  1.254926e+06     1.254926e+06  1.254926e+06  1.254926e+06   
mean   6.274625e+05     1.595409e+00  2.600214e+00  3.129635e-01   
std    3.622661e+05     1.246222e+00  2.319732e+00  4.636998e-01   
min    0.000000e+00     0.000000e+00  0.000000e+00  0.000000e+00   
25%    3.137312e+05     1.000000e+00  1.155128e+00  0.000000e+00   
50%    6.274625e+05     1.000000e+00  1.884102e+00  0.000000e+00   
75%    9.411938e+05     2.000000e+00  3.195736e+00  1.000000e+00   
max    1.254925e+06     6.000000e+00  2.931013e+01  1.000000e+00   

        holiday_Fri       weekday  trip_duration  
count  1.254926e+06  1.254926e+06   1.254926e+06  
mean   4.514975e-01  3.066662e+00   6.809164e+02  
std    4.976421e-01  1.963847e+00   4.321062e+02  
min    0.000000e+00  0.000000e+00   1.000000e+00  
25%    0.000000e+00  1.000000e+00   3.670000e+02  
50%    0.000000e+00  3.000000e+00   5.890000e+02  
75%    1.000000e+00  5.000000e+00   8.980000e+02  
max    1.000000e+00  6.000000e+00   4.645000e+03

In [7]:
model = LinearRegression()
cv = KFold(10)

In [14]:
taxi.head()

passenger_count      dist  holiday  holiday_Fri  weekday  trip_duration
0                1  1.497580        0            0        0            455
1                1  1.804374        1            1        6            663
2                1  1.484566        0            0        2            429
3                1  1.187842        1            1        5            435
4                6  1.098252        1            1        5            443

In [15]:
x = taxi.iloc[ : , : 4]
y = taxi.trip_duration

In [18]:
ls = cross_val_score(model, x, y, scoring = "r2", cv = cv)

In [19]:
ls.mean()

0.70442608545010965

- 10개의 테스트 모두 원래 모델의 성능과 유사하게 나옴.(테스트 10개의 평균값은 0.7044)
- 성능이 비슷하므로 과최적화가 되지 않았다고 판단할 수 있는 건가
- 만약 과최적화가 되지 않았다면, 기존 모델의 성능을 더 올려야 하는 것이 아닐까


In [5]:
#test_16 상태의 file로 cross_validation
taxi = pd.read_csv("latest_working_file.csv")
model = LinearRegression()
cv = KFold(10)

In [8]:
taxi.head()

passenger_count      dist  holiday  weekday  trip_duration
0                1  1.497580        1        1            455
1                1  1.804374        2        7            663
2                1  1.484566        1        3            429
3                1  1.187842        2        6            435
4                6  1.098252        2        6            443

In [9]:
x = taxi.iloc[ : , : 3]
y = taxi.trip_duration

In [10]:
ls = cross_val_score(model, x, y, scoring = "r2", cv = cv)

In [12]:
ls.mean()

0.72378372494607823